src/directive/rateLimit.ts

```diff
@@ -9,58 +12,135 @@ import {
 import { GraphQLError } from 'graphql';
 import { singleRedisClient } from '../redis';
 import { Context } from '../Context';
+import { logger } from '../logger';
+import { WATERCOOLER_ID } from '../common';
 
 export class CustomRateLimiterRedis extends RateLimiterRedis {
-  constructor(props) {
+  constructor(props: IRateLimiterRedisOptions) {
     super(props);
   }
 
   // Currently not doing any special actions/overrides
   // This was primarily introduced to make debugging easier by logging the details of rate limited queries/mutations after receiving a request
-  consume(key, pointsToConsume, options) {
+  consume(
+    key: string | number,
+    pointsToConsume?: number,
+    options?: { [key: string]: unknown },
+  ) {
     if (process.env.NODE_ENV === 'development') {
-      console.log(`[CONSUME] ${key} for ${pointsToConsume}`);
+      logger.info(`[CONSUME] ${key} for ${pointsToConsume}`);
     }
 
     return super.consume(key, pointsToConsume, options);
   }
 }
 
-const keyGenerator = (directiveArgs, source, args, context, info) =>
-  `${context.userId ?? context.trackingId}:${defaultKeyGenerator(
-    directiveArgs,
-    source,
-    args,
-    context,
-    info,
-  )}`;
+const keyGenerator: RateLimitKeyGenerator<Context> = (
+  directiveArgs,
+  source,
+  args,
+  context,
+  info,
+) => {
+  {
+    switch (info.fieldName) {
+      case 'createFreeformPost':
+      case 'submitExternalLink':
+      case 'sharePost':
+        return `${context.userId ?? context.trackingId}:createPost`;
+      case 'commentOnPost':
+      case 'commentOnComment':
+        return `${context.userId ?? context.trackingId}:createComment`;
+      default:
+        return `${context.userId ?? context.trackingId}:${defaultKeyGenerator(
+          directiveArgs,
+          source,
+          args,
+          context,
+          info,
+        )}`;
+    }
+  }
+};
 
 class RateLimitError extends GraphQLError {
   extensions = {};
   message = '';
-  constructor(msBeforeNextReset) {
+
+  constructor({
+    msBeforeNextReset,
+    message,
+  }: {
+    msBeforeNextReset?: number;
+    message?: string;
+  }) {
     const seconds = (msBeforeNextReset / 1000).toFixed(0);
-    const message = `Too many requests, please try again in ${seconds} seconds.`;
+    message = message ?? `Too many requests, please try again in ${seconds}s`;
     super(message);
 
     this.message = message;
     this.extensions = { code: 'RATE_LIMITED' };
   }
 }
 
-const onLimit = (resource) => {
-  throw new RateLimitError(resource.msBeforeNext);
+export const onLimit: RateLimitOnLimit<Context> = (
+  resource,
+  _,
+  __,
+  ___,
+  context,
+  info,
+) => {
+  switch (info.fieldName) {
+    case 'createFreeformPost':
+    case 'submitExternalLink':
+    case 'sharePost':
+      context.rateLimitCouner.add(1, { type: 'createPost' });
```

Comment:
Oh nvm see you add it here :)

Reply:
Yep 😄, didn't want to create a counter for each type. So instead I add some metadata to each count instead.